### PR TITLE
[BUG] handle } in line delimited json

### LINKED ISSUE #14390: BUG: Line delimited json is breaks if string includes `}`
In 0.19.0 exporting json with new lines breaks if a string contains a closing curly brace.

#### Example

```python
import pandas as pd
pd.DataFrame([{'a': 'foo}', 'b': 'bar'}, {'a': 'foo2', 'b': 'bar2'}]).to_json(orient='records', lines=True)

'{"a":"foo}"\n"b":"bar"},{"a":"foo2"\n"b":"bar2"}'
```

#### Expected Output
```python
'{"a":"foo}","b":"bar"}\n{"a":"foo2","b":"bar2"}'
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Darwin
OS-release: 16.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.0
nose: None
pip: 8.1.2
setuptools: 27.2.0
Cython: 0.24.1
numpy: 1.11.2
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: 1.1.0
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.4.0
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 3.6.4
bs4: 4.5.1
html5lib: 0.999
httplib2: 0.9.2
apiclient: 1.5.3
sqlalchemy: None
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: 2.42.0
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by joshowen**
- [x] closes #14390
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
can you add a benchmark for ``lines=True`` in ``asv_bench/benchmarks/packers.py`` (alongside the other json ones)

**Comment by joshowen** [comment]
Not sure why that test failed,, the only change was a line of whitespace

File: asv_bench/benchmarks/packers.py [hunk 1]
```diff
@@ -547,6 +547,36 @@ def remove(self, f):
...
 
 
+class packers_write_json_lines(object):
+    goal_time = 0.2
+
+    def setup(self):
+        self.f = '__test__.msg'
+        self.N = 100000
+        self.C = 5
+        self.index = date_range('20000101', periods=self.N, freq='H')
```
Comment (jreback) [inline asv_bench/benchmarks/packers.py]: looks like some things got repeated here

Comment (joshowen) [inline asv_bench/benchmarks/packers.py]: Looks like self.N/self.C are repeated in most of these classes.  Want me to clean them all up?

Comment (jreback) [inline asv_bench/benchmarks/packers.py]: sure that would be great (you can also make a common base class(s) if that helps as well)

Comment (jorisvandenbossche) [inline asv_bench/benchmarks/packers.py]: @joshowen you can leave it here as is. I cleaned this up in another PR (@jreback yes I know, I should merge that ...)

Comment (jreback) [inline asv_bench/benchmarks/packers.py]: ok that's fine (though @joshowen make sure your example doesn't have dups as this is new code)

Comment (jorisvandenbossche) [inline asv_bench/benchmarks/packers.py]: ah, yes, it is of course OK to remove the lines in this added code that you do not need for this benchmark

**Comment by jreback** [comment]
@joshowen can you also post a run for this benchmark

**Comment by joshowen** [comment]
@jreback is there an easy way to do that?  Or should I port the asv test to master and run/compare?

**Comment by jreback** [comment]
you *can* run asv if you want, otherwise just do it in ipython (before and after), e.g. something like

```
In [6]: df = DataFrame(dict([('float{0}'.format(i), np.random.randn(N)) for i in range(C)]), index=index).reset_index(drop=True)

In [7]: df.to_json('foo.json',orient='records',lines=True)

In [8]: %timeit df.to_json('foo.json',orient='records',lines=True)
1 loop, best of 3: 3.66 s per loop

In [9]: %timeit df.to_json('foo.json',orient='records')
10 loops, best of 3: 98.8 ms per loop

```

and looking at this, we have a BIG perf hit when ``lines=True``.

**Comment by jreback** [comment]
cc @aterrel

**Comment by jreback** [comment]
@joshowen can you open another issue about the perf 

```
In [15]: %prun df.to_json('foo.json',orient='records',lines=True)
         100058 function calls (100056 primitive calls) in 3.759 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    2.017    2.017    3.656    3.656 json.py:600(_convert_to_line_delimits)
        1    1.002    1.002    1.002    1.002 {method 'join' of 'str' objects}
        1    0.630    0.630    0.630    0.630 {numpy.core.multiarray.array}
        1    0.078    0.078    0.078    0.078 {pandas.json.dumps}
        1    0.012    0.012    0.012    0.012 {method 'write' of 'file' objects}
```

something odd going on here

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
Thanks @joshowen !

**Pull request closed without merging**